modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java

```diff
@@ -0,0 +1,299 @@
+                int catalogVersion = catalogService.latestCatalogVersion();
+
+                return getMvTableStorageFuture(parameters.causalityToken(), primaryReplicaId)
+                        .thenCompose(mvTableStorage -> awaitPrimaryReplicaFowNow(primaryReplicaId)
+                                .thenAccept(replicaMeta -> tryStartBuildIndexesForNewPrimaryReplica(
+                                        catalogVersion,
+                                        primaryReplicaId,
+                                        mvTableStorage,
+                                        replicaMeta
+                                ))
+                        );
+            } else {
+                stopBuildingIndexesIfPrimacyLost(primaryReplicaId);
+
+                return completedFuture(null);
+            }
+        });
+    }
+
+    private void tryStartBuildIndexesForNewPrimaryReplica(
+            int catalogVersion,
+            TablePartitionId primaryReplicaId,
+            MvTableStorage mvTableStorage,
+            ReplicaMeta replicaMeta
+    ) {
+        inBusyLock(busyLock, () -> {
+            if (isLeaseExpire(replicaMeta)) {
+                stopBuildingIndexesIfPrimacyLost(primaryReplicaId);
+
+                return;
+            }
+
+            // TODO: IGNITE-20530 We only need to get write-only indexes
+            for (CatalogIndexDescriptor indexDescriptor : catalogService.indexes(catalogVersion)) {
+                if (primaryReplicaId.tableId() == indexDescriptor.tableId()) {
+                    startBuildIndex(primaryReplicaId, indexDescriptor, mvTableStorage);
+                }
+            }
+        });
+    }
+
+    private void tryStartBuildIndex(
+            TablePartitionId primaryReplicaId,
+            CatalogIndexDescriptor indexDescriptor,
+            MvTableStorage mvTableStorage,
+            ReplicaMeta replicaMeta
+    ) {
+        inBusyLock(busyLock, () -> {
+            if (isLeaseExpire(replicaMeta)) {
+                stopBuildingIndexesIfPrimacyLost(primaryReplicaId);
+
+                return;
+            }
+
+            startBuildIndex(primaryReplicaId, indexDescriptor, mvTableStorage);
+        });
+    }
+
+    private void stopBuildingIndexesIfPrimacyLost(TablePartitionId replicaId) {
+        if (primaryReplicaIds.remove(replicaId)) {
+            // Primary replica is no longer current, we need to stop building indexes for it.
+            indexBuilder.stopBuildingIndexes(replicaId.tableId(), replicaId.partitionId());
+        }
+    }
+
+    private CompletableFuture<MvTableStorage> getMvTableStorageFuture(long causalityToken, TablePartitionId replicaId) {
+        return indexManager.getMvTableStorage(causalityToken, replicaId.tableId());
+    }
+
+    private CompletableFuture<ReplicaMeta> awaitPrimaryReplicaFowNow(TablePartitionId replicaId) {
+        return placementDriver
+                .awaitPrimaryReplica(replicaId, clock.now(), AWAIT_PRIMARY_REPLICA_TIMEOUT, SECONDS)
```

Comment:
> So the timeout constant is in seconds. Please reflect that in the name. By default all timeouts are defined in milliseconds, so this one leads to confusion

I don’t know about this, where can I look about this rule?

> If we wait indefinitely, then why do we split in into 10-seconds chunks?

As far as I can see, this is what is happening now, or I didn’t understand your proposal.

Reply:
> I don’t know about this, where can I look about this rule?

I don't think it's specified, but it is what we use commonly. It's like having the size in bytes by default, but then creating `long FOO_SIZE = 10` meaning 10 kilobytes, it's confusing.
> Or I didn’t understand your proposal.

It was a question. If we wait indefinitely, then why do we pass a 10 seconds timeout instead of passing an infinite timeout? This is literally what I asked, there was no proposal.

Reply:
> I don't think it's specified, but it is what we use commonly. It's like having the size in bytes by default, but then creating long FOO_SIZE = 10 meaning 10 kilobytes, it's confusing.

Commonly where? There are different options for the project, it may be more correct to add a unit of measurement of _MIILS, _SEC and etc?

> It was a question. If we wait indefinitely, then why do we pass a 10 seconds timeout instead of passing an infinite timeout? This is literally what I asked, there was no proposal.

Now I understand, I’ll try to wait forever.

Reply:
> Commonly where?

In the project. If there are other constants, that mean seconds without being named _SEC or something, they are just as confusing.
Again, I don't understand why we're arguing. Do you assume, that for every constant we should look for its usages to understand what it means? That would be an absurd claim, if we're not in tests.
I also understand that you copied the name without second thought. But the moment you did that, you became the author of the code. If you think that adding _SEC or _SECONDS will make the code worse, please explain why

Reply:
We discussed it in person, adding the postfix `_SEC`.